src/osp/universe/coordinates.h

```diff
@@ -24,316 +24,188 @@
-     */
-    constexpr CoordspaceTransform operator()(CoordspaceTransform const& in) const noexcept
+    else if (d > 0)
     {
-        CoordspaceTransform result;
-
-        result.m_expX = m_expX + in.m_expX;
-
-        int16_t expC1 = m_expC;
-        int16_t expC2 = m_expX + in.m_expC;
-
-        if (expC1 == expC2)
-        {
-            result.m_expC = expC1;
-            result.m_c = m_c + in.m_c;
-        }
-        else if (expC1 > expC2)
-        {
-            // expC1 too high, multiply into m_c
-            result.m_expC = expC2;
-            result.m_c = m_c * 2L * (1 << (expC1-expC2)) + in.m_c;
-
-        }
-        else if (expC1 < expC2)
-        {
-            // expC2 too high, multiply into in.m_c
-            result.m_expC = expC1;
-            result.m_c = m_c + in.m_c * 2L * (1 << (expC2-expC1));
-        }
-
-        return result;
+        c3 = f2.m_c * math::int_2pow<spaceint_t>(d) + f1.m_c;
+        m3 = f1.m_m;
+    }
+    else // if (d < 0)
+    {
+        c3 = f2.m_c + f1.m_c * math::int_2pow<spaceint_t>(-d);
+        m3 = f2.m_m + f1.m_n;
     }
 
-    int16_t m_expX{0};
-
-    Vec_t m_c{0};
-    int16_t m_expC{0};
-
-}; // struct CoordspaceTransform
-
-namespace transform
-{
-
-constexpr CoordspaceTransform scaled(
-        CoordspaceTransform x, int16_t from, int16_t to) noexcept
-{
-    return CoordspaceTransform{int16_t(x.m_expX + (from - to)), x.m_c, x.m_expC};
+    return {
+        .m_rotate   = f2.m_rotate * f1.m_rotate,
+        .m_c        = c3,
+        .m_n        = f1.m_n + f2.m_n,
+        .m_m        = m3
+    };
 }
 
-constexpr CoordspaceTransform child_to_parent(
-        Vector3g const childPos, int16_t childPrec, int16_t parentPrec) noexcept
+inline CoordTransformer coord_parent_to_child(
+        CoSpaceTransform const& parent, CoSpaceTransform const& child) noexcept
 {
-    int16_t exp = childPrec - parentPrec;
-    return CoordspaceTransform{exp, CoordspaceTransform::Vec_t(childPos), 0};
+    int const precisionDiff = child.m_precision - parent.m_precision;
+    return {
+        .m_rotate   = child.m_rotation.inverted(),
+        .m_c        = -child.m_position,
+        .m_n        = precisionDiff,
+        .m_m        = precisionDiff
+    };
 }
 
-constexpr CoordspaceTransform parent_to_child(
-        Vector3g const childPos, int16_t childPrec, int16_t parentPrec) noexcept
+inline CoordTransformer coord_child_to_parent(
+        CoSpaceTransform const& parent, CoSpaceTransform const& child) noexcept
 {
-    int16_t exp = parentPrec - childPrec;
-    // individual vector components is workaround for non-constexpr operator-()
-    return CoordspaceTransform{exp, CoordspaceTransform::Vec_t( -childPos.x(), -childPos.y(), -childPos.z()), exp};
+    int const precisionDiff = child.m_precision - parent.m_precision;
+    return {
+        .m_rotate   = child.m_rotation,
+        .m_c        = child.m_position,
+        .m_n        = -precisionDiff,
```

Comment:
minor, use std::negate() over simply "-". While they do the same thing, std::negate is much harder to miss when casually reading.